core/src/main/java/com/linecorp/armeria/server/VirtualHost.java

```diff
@@ -348,15 +354,46 @@ public boolean shutdownAccessLogWriterOnStop() {
     }
 
     /**
-     * Finds the {@link Service} whose {@link Router} matches the {@link RoutingContext}.
+     * Finds the {@link HttpService} whose {@link Router} matches the {@link RoutingContext}.
      *
-     * @param routingCtx a context to find the {@link Service}.
+     * @param routingCtx a context to find the {@link HttpService}.
      *
      * @return the {@link ServiceConfig} wrapped by a {@link Routed} if there's a match.
      *         {@link Routed#empty()} if there's no match.
      */
     public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx) {
-        return router.find(requireNonNull(routingCtx, "routingCtx"));
+        return findServiceConfig(routingCtx, false);
+    }
+
+    /**
+     * Finds the {@link HttpService} whose {@link Router} matches the {@link RoutingContext}.
+     *
+     * @param routingCtx a context to find the {@link HttpService}.
+     * @param useFallbackService whether to use the fallback {@link HttpService} when there is no match.
+     *                           If {@code true}, the returned {@link Routed} will always be present.
+     *
+     * @return the {@link ServiceConfig} wrapped by a {@link Routed} if there's a match.
+     *         {@link Routed#empty()} if there's no match.
+     */
+    public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx, boolean useFallbackService) {
+        final Routed<ServiceConfig> routed = router.find(requireNonNull(routingCtx, "routingCtx"));
+        if (routed.isPresent() || !useFallbackService) {
+            return routed;
+        }
+
+        if (routingCtx.isCorsPreflight()) {
+            // '403 Forbidden' is better for a CORS preflight request than '404 Not Found'.
+            routingCtx.deferStatusException(HttpStatusException.of(HttpStatus.FORBIDDEN));
+        } else if (routingCtx.deferredStatusException() == null) {
+            routingCtx.deferStatusException(HttpStatusException.of(HttpStatus.NOT_FOUND));
+        }
+
+        return Routed.of(fallbackServiceConfig.route(),
```

Comment:
Is there a reason not to add the fallback service to the router?

Reply:
I wasn't sure if it's a good idea to cache the resulting `Route`. Thoughts?

Reply:
Ah interesting point - I remember our path parsing cache only stores into the cache when the service succeeded, perhaps the routing cache should do something similar. But either way, for this PR that makes sense to prevent exploding cache - let's add a comment :)

Reply:
Added a comment :-)